ronn/geopm_report.7.ronn

```diff
@@ -0,0 +1,288 @@
+
+* `Report Extensions`:
+
+The report can be extended by agents, or by through the
+`--geopm-report-signals` option to `geopmlaunch` which corresponds to
+the `GEOPM_REPORT_SIGNALS` environment variable.  See the description
+for `--geopm-report-signals` in **geopmlaunch(1)** for more details.
+Signals in the `GEOPM_REPORT_SIGNALS` list will be added to individual
+regions, the epoch totals, and the application totals for each host.
+
+Agents can add keys to the report header, host section, or individual
+region sections using the `report_header()`, `report_host()` or
+`report_region()` methods respectively.  See **geopm::Agent(3)** for
+more information about the report extensions available to agents.
+
+
+## EXAMPLES
+
+Suppose we have a system with 2 sockets per node and 2 CPUs per packge
+(socket) (assume 1 CPU per core for simplicity).  The application
+places one process (MPI rank) on each CPU, and each rank executes the same
+application code, containing regions called "A" and "B".  The
+processes are assigned to cores as follows:
+
+      socket 0     |      socket 1
+  CPU 0  |  CPU 1  |  CPU 2  |  CPU 3
+---------+---------+---------+----------
+    1    |    2    |    3    |    4
+
+The following is an example of the region enter/exit and epoch events
+seen by the Controller.  This stream of events will be used to
+calculate the average runtime and count over ranks for regions and
+epoch, and the current region hash for sampling.
+
+The following is an example of the regions being entered and exited on
+each CPU.  The sample rate is 0.002 s; the region hash for the
+REGION_HASH sample at larger domains is shown on the right side of the
+table; "-" indicates the unmarked region.  Note that in a real report,
+the entry and exit times used to calculate the "runtime" may not
+line up exactly with the sample boundaries, and samples may not be
+exactly the same throughout the run.
+
+
+ time    CPU0   CPU1   CPU2   CPU3   ||  board   pkg0   pkg1
+-------------------------------------++----------------------
+0.001     -      -      -      -     ||
+0.002     A      -      -      A     ||   -       -      -
+0.003     A      A      -      A     ||
+0.004     A      A      A      A     ||   A       A      A
+0.005     -      A      A      A     ||
+0.006     B      A      A      A     ||   -       -      A
+0.007     B      B      A      B     ||
+0.008     B      B      B      B     ||   B       B      B
+0.009     B      B      B      B     ||
+0.010     B      B      B      B     ||   B       B      B
+0.011     -      -      -      -     ||
+0.012     finalize: report generated ||   -       -      -
+
+A subset of the report is shown below.  The "runtime" and "count"
+fields are averaged across the 4 CPUs.  The user extensions for
+"TIME@package" correspond to the sync runtime for each package.
+
+  Regions:
+    -
+    "name": "A",
+    "runtime": 0.00375,   // average of [0.002, 0.004, 0.004, 0.005]
```

Comment:
For the uninitiated calculation of the runtime here might be misleading since it looks like we are taking the sample times. I agree that there is no easy way to do that here but at least have a footnote emphasizing that runtime is not calculated based on sample.